tempesta_fw/sock.c

```diff
@@ -424,16 +488,20 @@ ss_do_close(struct sock *sk)
 	}
 }
 
-/*
- * This function is for internal Sync Sockets use only. It's called
- * under the socket lock taken by the kernel, and in the context of
- * the socket that is being closed.
+/**
+ * This function is for internal Sync Sockets use only. It's called under the
+ * socket lock taken by the kernel, and in the context of the socket that is
+ * being closed.
+ *
+ * This is unintentional connection closing, usually due to some data errors.
```

Comment:
Yes, we've lived for a long time with two callbacks `connection_drop()` and `connection_error()` which are basically just did the same thing (they actually called exactly the same code for either server and client connections). I commented purpose of the callbacks in `ss_hooks` definition. In short:

* connection_error - SS upcall letting upper layer to know about some errors in a TCP connection;

* connection_drop - is just a cleanup callback which is called when a TCP connection is already closed and we can do cleanup at Tempesta layer.

Please also note, that we have a consideration that TfwConnection lives longer than underlying TCP socket.

> `ss_droplink()` calls `connection_error()` callback now. That seems wrong. 

The comment explains the usage. "Intentional" means that we, Tempesta FW, intentionally closes connection. Basically, we don't care about other side intentions. If a client closes connection, then we just drop all associated resources with the connection, the same way as if there is some real connection error, e.g. TCP RST. The same for server: what's the difference why the server connection dies? We just move to failovering process regardless whether the server resets the connection or there were some error at TCP/IP layer. The different story is *intentional* closing, when Tempesta FW closes a connection - very different logic must be called in this case.